### PR TITLE
[supervisor] fixed flaky terminal tests

### PR DESCRIPTION
**Pull request by akosyakov**
get rid of time.sleep and instead wait till all content is read

### CODE REVIEW

File: components/supervisor/pkg/terminal/terminal_test.go [hunk 1]
```diff
@@ -49,17 +47,16 @@ func TestTerminals(t *testing.T) {
...
-				time.Sleep(test.Delay)
-			}
+			command := strings.Join(test.Stdin, " && ") + " && exit\r\n"
+			terminal.PTY.Write([]byte(command))
 
-			err = terminalService.Mux.Close(resp.Alias)
-			if err != nil {
-				t.Fatal(err)
-			}
+			<-done
```
Comment (akosyakov) [inline components/supervisor/pkg/terminal/terminal_test.go:59]: I tried to use `Process.Wait` but there were a race sometime, just waiting for end of file seems to be more reliable

File: components/supervisor/pkg/terminal/terminal_test.go [hunk 2]
```diff
@@ -49,17 +47,16 @@ func TestTerminals(t *testing.T) {
 				t.Fatal("no terminal")
 			}
 			stdoutOutput := bytes.NewBuffer(nil)
-			go io.Copy(stdoutOutput, terminal.Stdout.Listen())
+			done := make(chan struct{})
```
Comment (csweichel) [inline components/supervisor/pkg/terminal/terminal_test.go:50]: I think it's slightly simpler the other way 'round:

```
go func() {
  command := strings.Join(append(test.Stdin, "exit\r\n"), " && ")
  terminal.PTY.Write([]byte(command))
}()
io.Copy(stdoutOutput, terminal.Stdout.Listen())
```

**Comment by akosyakov** [comment]
@csweichel applied your suggestion

File: components/supervisor/pkg/terminal/terminal_test.go [hunk 3]
```diff
@@ -49,15 +46,12 @@ func TestTerminals(t *testing.T) {
...
-				time.Sleep(test.Delay)
-			}
-
-			err = terminalService.Mux.Close(resp.Alias)
-			if err != nil {
+			go func() {
+				command := strings.Join(append(test.Stdin, "sleep 1", "exit\r\n"), " && ")
+				terminal.PTY.Write([]byte(command))
+			}()
+			_, err = io.Copy(stdoutOutput, terminal.Stdout.Listen())
```
Comment (akosyakov) [inline components/supervisor/pkg/terminal/terminal_test.go:53]: It seems there is some kind of race: 
```
    terminal_test.go:61: unexpected output (-want +got):
          []string{
        - 	`echo "yarn" && echo "gp sync-done init" && echo "yarn --cwd theia-training watch" && history && sleep 1 && exit`,
        - 	"",
          	"",
          }
```

writer was closed, before delivering the complete content

**Comment by akosyakov** [comment]
superseded by https://github.com/gitpod-io/gitpod/pull/2020

**Pull request closed without merging**